### PR TITLE
Kafka example

### PR DESCRIPTION
**Pull request by jeqo**
Initial work to include a kafka-clients/kafka-streams example.

Includes Kafka Producer -> Streams -> Consumer flow.

Main challenge is to how to test it properly. We can wrap the producer as an Armeria server and kickout tests in the same way.

### CODE REVIEW

File: kafka/src/main/java/brave/example/MessagingTracingFactory.java [hunk 1]
```diff
@@ -0,0 +1,87 @@
...
+  /** Decides how to name and tag spans. By default they are named the same as the http method. */
+  static MessagingTracing create(String serviceName) {
+    return MessagingTracing.create(tracing((System.getProperty("brave.localServiceName", serviceName))));
+  }
+
+  /** Controls aspects of tracing such as the service name that shows up in the UI */
+  static Tracing tracing(String serviceName) {
+    return Tracing.newBuilder()
+        .localServiceName(serviceName)
+        .supportsJoin(Boolean.parseBoolean(System.getProperty("brave.supportsJoin", "true")))
```
Comment (jcchavezs) [inline kafka/src/main/java/brave/example/MessagingTracingFactory.java:33]: Maybe I am missing something but does `supportJoin` works for messaging?

File: kafka/src/main/java/brave/example/ProducerService.java [hunk 2]
```diff
@@ -0,0 +1,31 @@
...
+import java.util.Properties;
+
+public final class ProducerService {
+
+  public static void main(String[] args) {
+    final MessagingTracing messagingTracing = MessagingTracingFactory.create("producer-service");
+    final KafkaTracing kafkaTracing = KafkaTracing.newBuilder(messagingTracing).build();
+
+    final Properties configs = new Properties();
+    configs.put(ProducerConfig.BOOTSTRAP_SERVERS_CONFIG, "localhost:19092");
```
Comment (jcchavezs) [inline kafka/src/main/java/brave/example/ProducerService.java:20]: Will this work in containarized apps? Maybe making it configurable is "the voice"?

File: kafka/src/main/java/brave/example/MessagingTracingFactory.java (hunk 1 above)
Comment (jeqo) [reply kafka/src/main/java/brave/example/MessagingTracingFactory.java]: Agree, it doesn't make sense in this scenario.

File: kafka/src/main/java/brave/example/ProducerService.java (hunk 2 above)
Comment (jeqo) [reply kafka/src/main/java/brave/example/ProducerService.java]: right. adding System property `kafka.bootstrap-servers`

**Comment by jeqo** [comment]
@adriancole, components are rearranged into Frontend and Backend classes now, but it's still unclear to me how to do a couple of things:
- How to build images locally: haven't find the right command to build and test images locally. 
- How to run Kafka/Zookeeper as part of the tests. In the Cassandra use-case, a Daemon is run as back-end. With Kafka, we can run an embedded server, or adding containers to the compose config.

**Comment by adriancole** [comment]
Thanks for progressing this! Right now, all tests are executed via docker scripts. In the future we'll probably want to do a generic testcontainers thing so that they can also be run in java, but I'm concerned about classpaths so there's a chicken-egg on that. It would be best to test via the shell scripts in build-bin for now.

`build-bin/test armeria-kafka` should build and test the image. I would in this case actually make another docker-compose file in the build-bin directory for armeria-kafka that extends the other one similar to what we do in zipkin, extending frontend and backend as necessary (for the waits condition) https://github.com/openzipkin/zipkin/blob/master/docker/examples/docker-compose-kafka.yml

then, in build-bin/test change this line
```
build-bin/docker/docker_test_image ${docker_image} build-bin/docker-compose.test.yml
```

to something like..
```
case ${VERSION} in
  armeria-kafka* )
    docker_compose_file=build-bin/docker-compose-armeria-kafka.yml
    ;;
  * )
    docker_compose_file=build-bin/docker-compose.test.yml
    ;;
esac

build-bin/docker/docker_test_image ${docker_image} ${docker_compose_file}
```

don't forget to run `.github/generate_workflows` and check-in the new files. Otherwise GH actions won't run this.

**Review by adriancole — APPROVED** [review]
Should be good to go now!

Screen Shot 2020-12-24 at 7 07 37 PM

File: docker/bin/post-install-example-kafka [hunk 3]
```diff
@@ -0,0 +1,7 @@
+#!/bin/sh
+set -eux
+
+# Communication isn't HTTP
+sed -i '/endpoint=http:/d' start-frontend
+# Assign Kafka Bootstrap Servers from a variable
+sed -i '/-Dbrave.localServiceName/a-Dkafka.bootstrap-servers=${KAFKA_BOOTSTRAP_SERVERS:-kafka:9092} \\' start-frontend start-backend
```
Comment (adriancole) [inline docker/bin/post-install-example-kafka:7]: @jeqo I admit this would have been tricky to figure out, so just added it.

File: README.md [hunk 4]
```diff
@@ -41,6 +41,11 @@ Here are the example projects you can try:
   * Trace Instrumentation: [Armeria](https://armeria.dev/), [SLF4J](https://github.com/openzipkin/brave/tree/master/context/slf4j)
   * Trace Configuration: [Brave API](https://github.com/openzipkin/brave/tree/master/brave#setup) [Java](armeria/src/main/java/brave/example/HttpTracingFactory.java)
 
+* [armeria-kafka](armeria-kafka) `BRAVE_EXAMPLE=armeria-kafka docker-compose -f docker-compose.yml -f docker-compose-kafka.yml up`
```
Comment (adriancole) [inline README.md:44]: how to run is here, now. and will work post-merge

File: build-bin/docker-compose-kafka.test.yml [hunk 5]
```diff
@@ -0,0 +1,65 @@
...
+services:
+  sut:
+    container_name: sut
+    image: ghcr.io/openzipkin/alpine:3.12.3
+    entrypoint: /bin/sh
+    # Keep the container running until HEALTHCHECK passes
+    command: "-c \"sleep 5m\""
+    healthcheck:
+      # Return 0 when we can read back our trace ID
+      test: wget -qO- --spider http://zipkin:9411/api/v2/trace/cafebabecafebabe
```
Comment (adriancole) [inline build-bin/docker-compose-kafka.test.yml:14]: it occurs to me this doesn't actually test that backend sent a span (I checked and it does), but yeah.. we might want to change the test to include `| grep ...` where ... is something that verifies backend is the "local" servicename. we could install lazy jq possibly, but it is probably fine to just use grep. That, or we can count the spans.. ex |grep spanId|uniq or something

Comment (adriancole) [reply build-bin/docker-compose-kafka.test.yml:14]: cc @jcchavezs noting this isn't important to address here (possibly ever)

File: armeria-kafka/src/main/java/brave/example/Backend.java [hunk 6]
```diff
@@ -0,0 +1,94 @@
...
+    StringDeserializer valueDeserializer = new StringDeserializer();
+    Consumer<String, String> consumer = tracing.kafkaTracing
+        .consumer(new KafkaConsumer<>(consumerConfigs, keyDeserializer, valueDeserializer));
+
+    ExecutorService executorService = Executors.newFixedThreadPool(1);
+    executorService.submit(new ConsumerLoop(consumer, Collections.singleton("output")));
+
+    Server server = Server.builder()
+        .http(9000)
+        .service("/health", HealthCheckService.builder().build())
```
Comment (minwoox) [inline armeria-kafka/src/main/java/brave/example/Backend.java:62]: `HealthCheckService.of()`?

File: armeria-kafka/src/main/java/brave/example/Backend.java [hunk 7]
```diff
@@ -0,0 +1,94 @@
...
+    consumerConfigs.put(ConsumerConfig.BOOTSTRAP_SERVERS_CONFIG, kafkaBootstrapServers);
+    consumerConfigs.put(ConsumerConfig.GROUP_ID_CONFIG, "consumer-service");
+    consumerConfigs.put(ConsumerConfig.AUTO_OFFSET_RESET_CONFIG, "earliest");
+
+    StringDeserializer keyDeserializer = new StringDeserializer();
+    StringDeserializer valueDeserializer = new StringDeserializer();
+    Consumer<String, String> consumer = tracing.kafkaTracing
+        .consumer(new KafkaConsumer<>(consumerConfigs, keyDeserializer, valueDeserializer));
+
+    ExecutorService executorService = Executors.newFixedThreadPool(1);
```
Comment (minwoox) [inline armeria-kafka/src/main/java/brave/example/Backend.java:57]: nit: `Executors.newSingleThreadExecutor()`?

**Review by minwoox — APPROVED** [review]
Looks good! Thanks! 😄

**Comment by adriancole** [comment]
thanks for feedback @minwoox!

**Pull request merged**